### ISSUE TITLE
Feature request - would it be possible to auto enrol users into courses and category/system roles

### ISSUE THREAD

**Issue by swinster** [microsoft/o365-moodle#1178]
On AD sync, it would be great if users in different OU's could be automatically enrolled on courses in specific categories. For example, Teachers would be enrolled on Health and Safety, GDPR, Conduct etc, and Students could be enrolled on Conduct, Plagiarism, Student Life, etc. This would means that new users added to AD are auto enrolled on their relevant standard courses. 

We actually use Moodle in a workplace and external training context so things like sales, engineering, support, general are all relevant here too. 

In addition, we use roles in various categories to prevent users from seeing courses/categories that they should not (such as not employees cannot see internal training, but equally students could not see Teaching staff material.

**Comment by swinster** [comment]
Possible overlap with https://github.com/microsoft/o365-moodle/issues/210 and https://github.com/microsoft/o365-moodle/issues/1174

**Comment by swinster** [comment]
Potentially, a bit of https://github.com/microsoft/o365-moodle/issues/1095 as well. 

Basically, we need a simple way not just to import users, but to enroll them to the correct course(s). Whether this is done through direct enrolment at sync, or into a cohort or some other way I don't know, but as it stands, this all seem like a manual task. FOr a few hundred, it a PITA, for 10's thousands, honestly I don't know how they would do that.

**Comment by nenorojas** [comment]
@swinster There are no plans to create a new plugin to manage enrolment in Moodle as part of the o365 set. We recommend to try this plugin that allows enrolment based in profile fields: https://moodle.org/plugins/enrol_attributes
There's a new core feature for Moodle Workplace (different version of Moodle available via Moodle Partners) called "Dynamic Rules" https://docs.moodle.org/38/en/Dynamic_rules, that allows enrolment based in profile fields and more. So there's a chance it may be part of future Moodle versions.
On a separated note, allowing extra fields to be mapped is definitely something that will be checked:
https://github.com/microsoft/o365-moodle/issues/210
https://github.com/microsoft/o365-moodle/issues/1104

I'm closing this ticket.

**Comment by swinster** [comment]
@nenorojas many thanks for this. FWIW, we worked around the issue using something a litel more convoluted, but gave us another benefit. I should have posted an update. For posterity, I will add it here just in case anyone has a similar issue in the future:

We used two plugins:

* Profile field based cohort membership
* Cohort role synchronization

The first use a hidden file (much like above - organisation name in our case, which obviously has to be set correctly in AD) to add the users into a System context role.

The second syncs that role with a cohort.

We then use the native cohort enrolment method to enrol users.

The benefit of creating a system role as well, is that we have both internal and external users using Moodle. We can then setup visible Course Categories with specific permissions that allow internal users to view them, whilst external user can't. 

I did look at Moodle Workplace and this is definitely something that is on the radar, how after speaking with our Moodle partner, it was suggested that this was not the way to go for the time being. However, there are other feature in Workplace that I am keen to see.

Thanks for your update.
